### PR TITLE
executor: control Chunk size for Joiners

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Control the number of rows in chunks returned by `joiner` and join-like executors.
Following up #9452, this PR is a subtask of #9166.

### What is changed and how it works?
For `MergeJoin`, `HashJoin` and `IndexLookupJoin`, support `requiredRows` from parents and push `requiredRows` down to its outer child if it is outer join.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by qw4990** [comment]
@lysu @zz-jason PTAL~

File: executor/merge_join.go [hunk 1]
```diff
@@ -340,7 +340,7 @@ func (e *MergeJoinExec) joinToChunk(ctx context.Context, chk *chunk.Chunk) (hasM
 			e.innerIter4Row.Begin()
 		}
 
-		if chk.NumRows() >= e.maxChunkSize {
+		if chk.IsFull() {
```
[the remark this answers is not in the document]
Comment (qw4990) [reply executor/merge_join.go]: Yes you are right.
It has been updated.

File: executor/merge_join.go [hunk 2]
```diff
@@ -377,7 +377,13 @@ func (e *MergeJoinExec) fetchNextInnerRows() (err error) {
 // fetchNextOuterRows fetches the next Chunk of outer table. Rows in a Chunk
 // may not all belong to the same join key, but are guaranteed to be sorted
 // according to the join key.
-func (e *MergeJoinExec) fetchNextOuterRows(ctx context.Context) (err error) {
+func (e *MergeJoinExec) fetchNextOuterRows(ctx context.Context, requiredRows int) (err error) {
+	// It's hard to calculate selectivity if there is any filter or it's inner join,
+	// so we just push the requiredRows down when it's outer join and has no filter.
+	if IsOuterJoiner(e.joiner) && len(e.outerTable.filter) == 0 {
```
Comment (zz-jason) [inline executor/merge_join.go]: the property "is outer join" only need to be calculated once, we can use a variable to store that result

File: executor/index_lookup_join.go [hunk 3]
```diff
@@ -360,9 +372,14 @@ func (ow *outerWorker) buildTask(ctx context.Context) (*lookUpJoinTask, error) {
 	task.memTracker.AttachTo(ow.parentMemTracker)
 
 	ow.increaseBatchSize()
+	task.outerResult.SetRequiredRows(ow.batchSize, ow.maxBatchSize)
```
Comment (lzmhhh123) [inline executor/index_lookup_join.go]: How about
```suggestion
	else {
		task.outerResult.SetRequiredRows(ow.batchSize, ow.maxBatchSize)
	}
```
to reduce a function call.

File: executor/index_lookup_join.go [hunk 4]
```diff
@@ -103,6 +107,8 @@ type lookUpJoinTask struct {
 type outerWorker struct {
 	outerCtx
 
+	lookup *IndexLookUpJoin
```
Comment (lzmhhh123) [inline executor/index_lookup_join.go]: Why not only save the `isOuterJoin` and move it into the `outerCtx`?

Comment (qw4990) [reply executor/index_lookup_join.go]: updated.

File: executor/index_lookup_join.go (hunk 3 above)
Comment (qw4990) [reply executor/index_lookup_join.go]: updated.

File: executor/merge_join.go (hunk 2 above)
Comment (qw4990) [reply executor/merge_join.go]: It's updated.

File: executor/index_lookup_join.go [hunk 5]
```diff
@@ -165,6 +171,8 @@ func (e *IndexLookUpJoin) Open(ctx context.Context) error {
 }
 
 func (e *IndexLookUpJoin) startWorkers(ctx context.Context) {
+	e.isOuterJoin = IsOuterJoiner(e.joiner)
+	e.requiredRows = int64(e.maxChunkSize)
```
[the remark this answers is not in the document]
Comment (qw4990) [reply executor/index_lookup_join.go]: It has been removed.

File: executor/joiner.go [hunk 6]
```diff
@@ -460,3 +460,13 @@ func (j *innerJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *ch
 
 func (j *innerJoiner) onMissMatch(_ bool, outer chunk.Row, chk *chunk.Chunk) {
 }
+
+// IsOuterJoiner returns if this joiner is a outer joiner
+func IsOuterJoiner(j joiner) bool {
```
[the remark this answers is not in the document]
Comment (qw4990) [reply executor/joiner.go]: Updated.

File: executor/index_lookup_join.go (hunk 4 above)
Comment (qw4990) [reply executor/index_lookup_join.go]: `outerWorker` also needs `IndexLookUpJoin.requiredRows`, so it maybe more clearer than put these two variables `RequiredRows` and `IsOuterJoin` into `outerCtx`.

**Comment by qw4990** [comment]
All comments have been addressed, PTAL @zz-jason @lzmhhh123

File: executor/joiner.go [hunk 7]
```diff
@@ -460,3 +460,13 @@ func (j *innerJoiner) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *ch
 
 func (j *innerJoiner) onMissMatch(_ bool, outer chunk.Row, chk *chunk.Chunk) {
 }
+
+// IsOuterJoiner returns if this joiner is a outer joiner
+func IsOuterJoiner(j plannercore.JoinType) bool {
```
Comment (zz-jason) [inline executor/joiner.go]: how about bind this function to `JoinType`? thus we can call `joinType.IsOuterJoin()` to achieve this functionality.

Comment (qw4990) [reply executor/joiner.go]: updated.

**Pull request merged**